### PR TITLE
implement validation for GatewayClass resource

### PR DESCRIPTION
**Pull request by hbagdi**
This currently comprises only of an immutability check for
`spec.controller` field. This can expanded in future.

**What type of PR is this?**

/kind feature
**What this PR does / why we need it**:
This validation is a pre-requisite for #189. Once this is merged in, we can start enforcing this using a validation webhook.
**Which issue(s) this PR fixes**:

None
**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: apis/v1alpha2/validation/validation_test.go [hunk 1]
```diff
@@ -424,3 +426,68 @@ func portNumberPtr(p int) *gatewayv1a2.PortNumber {
...
 }
+
+func TestValidateGatewayClassUpdate(t *testing.T) {
+	type args struct {
+		oldClass *gatewayv1a2.GatewayClass
+		newClass *gatewayv1a2.GatewayClass
+	}
+	tests := []struct {
+		name string
+		args args
```
Comment (robscott) [inline apis/v1alpha2/validation/validation_test.go:437]: This is a tiny nit/question - any reason not to just directly embed oldClass and newClass here? At first I thought this might be the specific type that `ValidateGatewayClassUpdate` takes, but it looks like it just takes 2 individual GatewayClasses. This doesn't have to block this PR though.

**Review by robscott — COMMENTED** [review]
Thanks for the work on this!

File: apis/v1alpha2/validation/validation_test.go (hunk 1 above)
Comment (hbagdi) [reply apis/v1alpha2/validation/validation_test.go:437]: I use https://github.com/cweill/gotests to auto-generate these tables.

**Comment by jpeach** [comment]
xref https://github.com/kubernetes/kubernetes/issues/65973

File: apis/v1alpha2/validation/validation.go [hunk 2]
```diff
@@ -122,3 +122,18 @@ func validateHTTPRouteUniqueFilters(rules []gatewayv1a2.HTTPRouteRule, path *fie
 
 	return errs
 }
+
+// ValidateGatewayClassUpdate validates class according to the Gateway API specification.
```
Comment (jpeach) [inline apis/v1alpha2/validation/validation.go:126]:
```suggestion
// ValidateGatewayClassUpdate validates an update to oldClass according to the Gateway API specification.
```

File: apis/v1alpha2/validation/validation_test.go [hunk 3]
```diff
@@ -424,3 +426,68 @@ func portNumberPtr(p int) *gatewayv1a2.PortNumber {
...
+			},
+			want: field.ErrorList{
+				{
+					Type:     field.ErrorTypeInvalid,
+					Field:    "spec.controller",
+					Detail:   "cannot update an immutable field",
+					BadValue: "bar",
+				},
+			},
+		},
```
Comment (jpeach) [inline apis/v1alpha2/validation/validation_test.go:483]: Suggest adding test cases where one (or both parameter is nil).

**Comment by hbagdi** [comment]
@robscott @jpeach Should we be implementing this validation for v1alpha1 and v1alpha2 both?

**Comment by hbagdi** [comment]
I made the mistake of pushing to this repo and creating this PR. This results in a protected branch and I can't force-push a protected branch so I'm going to close this. I've opened up https://github.com/kubernetes-sigs/gateway-api/pull/663 to continue the work here. Apologies for the inconvenience.

/close
